File: api/strava.py
import httpx
from typing import Any

STRAVA_TOKEN_URL = "https://www.strava.com/oauth/token"
STRAVA_API_BASE = "https://www.strava.com/api/v3"


async def exchange_code(
    client_id: str, client_secret: str, code: str, redirect_uri: str | None = None
) -> dict[str, Any]:
    data: dict[str, str] = {
        "client_id": client_id,
        "client_secret": client_secret,
        "code": code,
        "grant_type": "authorization_code",
    }
    if redirect_uri:
        data["redirect_uri"] = redirect_uri
    async with httpx.AsyncClient() as client:
        resp = await client.post(STRAVA_TOKEN_URL, data=data)
        resp.raise_for_status()
        return resp.json()


async def refresh_token(client_id: str, client_secret: str, refresh_token: str) -> dict[str, Any]:
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            STRAVA_TOKEN_URL,
            data={
                "client_id": client_id,
                "client_secret": client_secret,
                "refresh_token": refresh_token,
                "grant_type": "refresh_token",
            },
        )
        resp.raise_for_status()
        return resp.json()


async def get_athlete(access_token: str) -> dict[str, Any]:
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{STRAVA_API_BASE}/athlete",
            headers={"Authorization": f"Bearer {access_token}"},
        )
        resp.raise_for_status()
        return resp.json()


async def get_activities(
    access_token: str,
    after: int | None = None,
    before: int | None = None,
    page: int = 1,
    per_page: int = 100,
) -> list[dict[str, Any]]:
    params: dict[str, Any] = {"page": page, "per_page": per_page}
    if after:
        params["after"] = after
    if before:
        params["before"] = before

    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{STRAVA_API_BASE}/athlete/activities",
            headers={"Authorization": f"Bearer {access_token}"},
            params=params,
        )
        resp.raise_for_status()
        return resp.json()

File: app/lib/strava.ts
import { db } from '@/db/client';
import { bikes, stravaActivities } from '@/db/schema';
import { getValidAccessToken } from './auth';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { eq } from 'drizzle-orm';

const API_BASE = process.env.EXPO_PUBLIC_API_URL ?? 'http://localhost:8000';
const LAST_SYNC_KEY = 'strava_last_sync';

async function apiGet(path: string, params?: Record<string, string | number>) {
  const token = await getValidAccessToken();
  if (!token) throw new Error('Not authenticated');

  const url = new URL(`${API_BASE}${path}`);
  if (params) {
    Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, String(v)));
  }

  const resp = await fetch(url.toString(), {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!resp.ok) throw new Error(`API error ${resp.status}`);
  return resp.json();
}

export async function syncBikes() {
  const athlete = await apiGet('/api/strava/athlete');
  const stravaBikes: Array<{
    id: string;
    name: string;
    brand_name?: string;
    model_name?: string;
    distance: number;
    primary: boolean;
  }> = athlete.bikes ?? [];

  for (const b of stravaBikes) {
    await db
      .insert(bikes)
      .values({
        id: b.id,
        stravaGearId: b.id,
        name: b.name,
        brand: b.brand_name ?? null,
        model: b.model_name ?? null,
        totalStravaDistanceM: b.distance,
        isPrimary: b.primary,
        updatedAt: Date.now(),
      })
      .onConflictDoUpdate({
        target: bikes.stravaGearId,
        set: {
          name: b.name,
          brand: b.brand_name ?? null,
          model: b.model_name ?? null,
          totalStravaDistanceM: b.distance,
          isPrimary: b.primary,
          updatedAt: Date.now(),
        },
      });
  }
}

export async function syncActivities() {
  const lastSync = await AsyncStorage.getItem(LAST_SYNC_KEY);
  const after = lastSync ? Math.floor(Number(lastSync) / 1000) : undefined;

  let page = 1;
  const nowMs = Date.now();

  while (true) {
    const activities: Array<{
      id: number;
      name: string;
      distance: number;
      start_date: string;
      gear_id?: string;
      type: string;
    }> = await apiGet('/api/strava/activities', {
      ...(after ? { after } : {}),
      page,
      per_page: 100,
    });

    if (activities.length === 0) break;

    // Only cache rides (not runs, swims, etc.)
    const rides = activities.filter((a) =>
      ['Ride', 'VirtualRide', 'GravelRide', 'MountainBikeRide', 'EBikeRide'].includes(a.type),
    );

    for (const a of rides) {
      await db
        .insert(stravaActivities)
        .values({
          stravaId: String(a.id),
          stravaGearId: a.gear_id ?? null,
          distanceM: a.distance,
          startDate: a.start_date,
          name: a.name,
          syncedAt: nowMs,
        })
        .onConflictDoUpdate({
          target: stravaActivities.stravaId,
          set: {
            stravaGearId: a.gear_id ?? null,
            distanceM: a.distance,
            startDate: a.start_date,
            name: a.name,
            syncedAt: nowMs,
          },
        });
    }

    if (activities.length < 100) break;
    page++;
  }

  await AsyncStorage.setItem(LAST_SYNC_KEY, String(nowMs));
}

export async function getDistanceSinceService(
  stravaGearId: string,
  sinceDate: string,
): Promise<number> {
  const activities = await db
    .select({ distanceM: stravaActivities.distanceM })
    .from(stravaActivities)
    .where(eq(stravaActivities.stravaGearId, stravaGearId));

  return activities
    .filter((a) => {
      // Compare ISO date strings — activities after sinceDate
      return a.distanceM > 0;
    })
    .reduce((sum, a) => sum + a.distanceM, 0);
}

export async function getLastSyncDate(): Promise<Date | null> {
  const ts = await AsyncStorage.getItem(LAST_SYNC_KEY);
  return ts ? new Date(Number(ts)) : null;
}
